### ISSUE TITLE
z

### ISSUE THREAD

**Issue by xyzxfwen** [xyzxfwen/z#1]
"""
import os
import numpy as np
import paddle
from paddle.io import DataLoader
from paddle.vision.transforms import ToTensor
from PIL import Image
import random

# 数据集路径
data_root_path = "D:/植物分类/下载飞桨数据集/cls_chinese_medicine/"
train_image_list_path = os.path.join(data_root_path, 'train.txt')
val_image_list_path = os.path.join(data_root_path, 'val.txt')
label_file_path = os.path.join(data_root_path, 'label.txt')

# 加载标签
def load_labels(label_file):
    labels = {}
    with open(label_file, 'r') as f:
        for line in f:
            line = line.strip().split()
            if len(line) < 2:
                print(f"跳过无效行: {line}")
                continue
            try:
                labels[line[0]] = int(line[1])
            except ValueError:
                print(f"无法将标签转换为整数: {line[1]}，该行被跳过。")
    return labels

labels_dict = load_labels(label_file_path)

# 数据清洗与加载数据
def load_and_clean_data(image_list_path):
    images = []
    labels = []
    paths_set = set()

    with open(image_list_path, 'r') as f:
        for line in f:
            line = line.strip().split()
            if not line:
                continue
            img_path = os.path.join(data_root_path, line[0])

            if img_path in paths_set:
                continue
            paths_set.add(img_path)

            if not os.path.exists(img_path):
                print(f"图像文件不存在：{img_path}")
                continue

            try:
                img = Image.open(img_path).convert('RGB').resize((64, 64))
                img = ToTensor()(img)
                images.append(np.array(img))
                labels.append(labels_dict[line[0]])
            except Exception as e:
                print(f"无法加载图像 {img_path}: {e}")

    return np.array(images), np.array(labels)

# 加载训练数据
all_train_data, all_train_labels = load_and_clean_data(train_image_list_path)

# 划分训练集和测试集
num_train = len(all_train_data)
indices = list(range(num_train))
random.shuffle(indices)

split = int(0.9 * num_train)
train_indices = indices[:split]
test_indices = indices[split:]

train_data = all_train_data[train_indices]
train_labels = all_train_labels[train_indices]
test_data = all_train_data[test_indices]
test_labels = all_train_labels[test_indices]

# 加载验证数据
val_data, val_labels = load_and_clean_data(val_image_list_path)

# 数据预处理
def preprocess(data):
    return data.astype("float32") / 255.0

train_data = preprocess(train_data)
test_data = preprocess(test_data)
val_data = preprocess(val_data)

# 输出各个数据集的信息
def print_dataset_info(data, labels, dataset_name):
    num_images = len(data)
    unique_labels = np.unique(labels)
    num_classes = len(unique_labels)
    print(f"{dataset_name} - 图片数量: {num_images}, 类别数量: {num_classes}, 类别: {unique_labels.tolist()}")

print_dataset_info(train_data, train_labels, "训练集")
print_dataset_info(test_data, test_labels, "测试集")
print_dataset_info(val_data, val_labels, "验证集")

# 定义数据集类
class PlantDataset(paddle.io.Dataset):
    def __init__(self, data, labels):
        self.data = data
        self.labels = labels

    def __getitem__(self, idx):
        image = self.data[idx]
        label = self.labels[idx]
        return image, label

    def __len__(self):
        return len(self.data)

# 创建数据集
train_dataset = PlantDataset(train_data, train_labels)
test_dataset = PlantDataset(test_data, test_labels)
val_dataset = PlantDataset(val_data, val_labels)

# 定义DataLoader
BATCH_SIZE = 32
train_loader = DataLoader(train_dataset, batch_size=BATCH_SIZE, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=BATCH_SIZE)
val_loader = DataLoader(val_dataset, batch_size=BATCH_SIZE)

# 启用静态图模式
paddle.enable_static()

# 定义数据
image = paddle.static.data(name="image", shape=[None, 3, 64, 64], dtype="float32")
label = paddle.static.data(name="label", shape=[None, 1], dtype="int64")

# 构建CNN模型
class ConvolutionNeuralNetwork(paddle.nn.Layer):
    def __init__(self):
        super(ConvolutionNeuralNetwork, self).__init__()
        self.conv1 = paddle.nn.Conv2D(in_channels=3, out_channels=32, kernel_size=3, padding=1)
        self.max_pool1 = paddle.nn.MaxPool2D(kernel_size=2, stride=2)
        self.conv2 = paddle.nn.Conv2D(in_channels=32, out_channels=64, kernel_size=3, padding=1)
        self.max_pool2 = paddle.nn.MaxPool2D(kernel_size=2, stride=2)
        self.flatten = paddle.nn.Flatten()
        self.fc1 = paddle.nn.Linear(in_features=64 * 16 * 16, out_features=128)
        self.fc2 = paddle.nn.Linear(in_features=128, out_features=len(labels_dict))

    def forward(self, x):
        x = self.max_pool1(paddle.nn.functional.relu(self.conv1(x)))
        x = self.max_pool2(paddle.nn.functional.relu(self.conv2(x)))
        x = self.flatten(x)
        x = self.fc1(x)
        x = self.fc2(x)
        return x

# 创建CNN模型
predict = ConvolutionNeuralNetwork()

# 损失函数与训练过程
cost = paddle.nn.functional.cross_entropy(input=predict(image), label=label)
avg_cost = paddle.mean(cost)
accuracy = paddle.metric.accuracy(input=predict(image), label=label)
optimizer = paddle.optimizer.Adam(learning_rate=0.001)
optimizer.minimize(avg_cost)

# 执行器
place = paddle.CUDAPlace(0)
exe = paddle.static.Executor(place)
exe.run(paddle.static.default_startup_program())

# 训练模型
for epoch in range(40):
    for batch_id, data in enumerate(train_loader()):
        images = np.array([x[0] for x in data]).astype('float32')
        labels = np.array([x[1] for x in data]).reshape(-1, 1).astype('int64')
        loss, acc = exe.run(paddle.static.default_main_program(),
                            feed={"image": images, "label": labels},
                            fetch_list=[avg_cost, accuracy])
        if batch_id % 20 == 0:
            print("Epoch %d, Batch %d, Cost %.5f, Accuracy %.5f" % (epoch, batch_id, loss[0], acc[0]))

# 保存模型
model_save_dir = "model/plants/"
if not os.path.exists(model_save_dir):
    os.makedirs(model_save_dir)
paddle.save(model=predict, path=model_save_dir)

print("训练模型保存完毕！")

# 预测功能
def load_img(path):
    img = Image.open(path).resize((64, 64)).convert('RGB')
    img = ToTensor()(img)
    img = img.numpy() / 255.0
    return img

# 测试图像路径
test_img = "./data/test_image.png"
infer_imgs = load_img(test_img)[np.newaxis, :, :, :]

# 定义执行器
place = paddle.CPUPlace()
exe = paddle.static.Executor(place)
model_save_dir = "model/plants/"

# 模型载入
[inference_program, feed_target_names, fetch_targets] = paddle.static.load_inference_model(model_save_dir, exe)

# 执行预测
results = exe.run(inference_program,
                  feed={feed_target_names[0]: infer_imgs},
                  fetch_list=fetch_targets)

print("预测结果：", np.argmax(results[0]))

# 显示待预测的图像
img = Image.open(test_img)
img.show()
"""
